### PR TITLE
feat: add support for GitLab repositories

### PR DESCRIPTION
**Pull request by tigermarques**
**What**:

This change will introduce two new configurations:
* `repoType` represents the type of repository and is a choice between `github` and `gitlab` (default is `github`)
* `repoHost` represents where the repository is hosted, and it's default is `https://github.com` when `repoType` is `github`, and `https://gitlab.com` when `repoType` is `gitlab`.

**Why**:

These new configurations will allow to populate the contributors table with users that belong to registries other than https://github.com, which is currently the only one supported.

**How**:

By adding the two new configurations in the `init` step, and later using them while adding new contributors or generating the contributor list.

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Ready to be merged 
- [x] Added myself to contributors table

### CODE REVIEW

**Comment by tigermarques** [comment]
Hi @machour , do you want to take a look. I've done a bit of refactoring as well, in order to move all interactions with github to its own folder, to allow for a better abstraction for gitlab.

I made sure the existing unit tests kept working, and added my own tests to most of the code (but not all of it). I also tested it in a project I have on a private gitlab server, and it works like a charm :)

I'd appreciate your input in order to change whatever necessary

**Comment by tigermarques** [comment]
Sure I'll look into it. I'm used to having eslint shout at me about that. I'll change the indents to two spaces

**Comment by tigermarques** [comment]
Fixed, and I've added two new rules to the `eslintConfig` section in `package.json`

```
{
    ...
    "no-mixed-spaces-and-tabs": "error",
    "indent": ["error", 2]
}
```

This enforces what you asked me to change so that eslint shouts at me. Do you want me to push this as well? For the time being, it is only on my own `package.json`

**Comment by tigermarques** [comment]
@machour , I found these two npm packages that might be an option to replace our interactions with [github](https://www.npmjs.com/package/github-api) and [gitlab](https://www.npmjs.com/package/gitlab). What do you think? Maybe I could consider it for a future PR after you have decided whether to keep this feature or not.

**Comment by tigermarques** [comment]
@machour, have you had a chance to look at the code?

**Comment by machour** [comment]
@tigermarques I will try my best to do that today (I need to create a gitlab account and test everything so this might take me a while).

Could you fix the little merge conflict?

**Comment by tigermarques** [comment]
Merge conflict fixed. Added the check for >= 400 status codes for both github and gitlab.

**Comment by tigermarques** [comment]
If you want to user my user when testing in gitlab, fell free. It's [jmarques](https://gitlab.com/jmarques).

File: src/repo/gitlab.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      return request
+        .get({
+          url: `${hostname}/api/v4/projects/${project.id}/repository/contributors`,
+          headers: {
+            'User-Agent': 'request',
+          },
+        })
+        .then(newRes => {
+          const contributors = JSON.parse(newRes.body)
+					if (newRes.statusCode >= 400) {
```
Comment (machour) [inline src/repo/gitlab.js]: Some tabs leftovers

File: src/repo/gitlab.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		        throw new Error(contributors.message)
+		      }
+          return contributors.map(item => item.name)
```
Comment (machour) [inline src/repo/gitlab.js]: So sad that GitLab doesn't return the username.
I tested by changing my name in the `.allcontributorsrc` file to "Mehdi Acho".

Running the check command gives:
```
Missing contributors in .all-contributorsrc:
    Mehdi Achour
```

It also should have mentioned the unknown contributor found (Acho).
This happens around line 108 in src/cli.js

File: src/cli.js [hunk 3]
```diff
@@ -74,25 +75,26 @@ function addContribution(argv) {
 function checkContributors(argv) {
   const configData = util.configFile.readConfig(argv.config)
 
-  return util
-    .check(configData.projectOwner, configData.projectName)
+  return repo
+    .getContributors(configData.projectOwner, configData.projectName, configData.repoType, configData.repoHost)
     .then(ghContributors => {
```
Comment (machour) [inline src/cli.js]: `ghContributors` used to stand for GitHub Contributors.
Could you rename?

File: src/cli.js [hunk 4]
```diff
@@ -74,25 +75,26 @@ function addContribution(argv) {
...
-        contributor => contributor.login,
+        contributor => contributor[checkKey],
       )
 
       const missingInConfig = ghContributors.filter(
-        login => !knownContributors.includes(login),
+        key => !knownContributors.includes(key),
       )
-      const missingFromGithub = knownContributors.filter(login => {
+      const missingFromGithub = knownContributors.filter(key => {
```
Comment (machour) [inline src/cli.js]: Same here, this needs to be renamed

File: src/repo/gitlab.js (hunk 2 above)
Comment (machour) [reply src/repo/gitlab.js]: This comment got hidden as lines were changed. Don't forget to check it ;)

**Comment by machour** [comment]
@tigermarques just making sure you've read this comment, which is blocking the approval:

screen shot 2018-02-23 at 12 26 32 pm

File: src/repo/gitlab.js (hunk 2 above)
Comment (tigermarques) [reply src/repo/gitlab.js]: You're right I had missed it! So basically the error is that it detects the Gitlab user is not in the file, but does not detect that the contributor in the file is not in Gitlab. Is that it?

Comment (machour) [reply src/repo/gitlab.js]: exactly "is not in the project's contributors as returned by Gitlab" to be precise

**Comment by tigermarques** [comment]
I've checked the comment now, and here is a test I produced:

1. create a repo and commit with my user
2. generate `.all-contributorsrc`
3. add myself as contributor with `doc` contribution
4. run check - does not detect anything as expected
5. manually edit my name in the `.all-contributorsrc`
6. run check:
    * Detects one entry in `Missing contributors in .all-contributorsrc` as expected
    * Does not detect any `Unknown contributors` because my contributions do not include neither `code` nor `test`
7. manually edit my contributions in the `.all-contributorsrc` to add `code`
8. run check - detects both differences now as expected

Can you check if by any chance your contributions are of types that don't include `code` or `test`. The code there is very explicit:

```
const missingFromRepo = knownContributors.filter(key => {
    return (
        !repoContributors.includes(key) &&
        (knownContributions[key].includes('code') ||
          knownContributions[key].includes('test'))
    )
})
```

**Comment by machour** [comment]
Gotcha, I didn't have a `code` or `test` contribution type. Added `code` and verified that it works like a charm.

reaaaaaaally minor thingy: when the repo haven't got any contribution yet (just created), if you run check you'll get `404 Not found` printed without any explanations. This is because the gitlab api for contributors returns 404 in that case. 

Might be better to catch it and output "no contributors found on GitLab" 🤔

**Comment by tigermarques** [comment]
Good catch. I was returning the default message that Gitlab sent.
For the sake of consistency, I added the same message for both github and gitlab (difference in the repo name only)

**Comment by machour** [comment]
Also, since this is a big change, and people are used to this lib working with GitHub only, it may be worth hinting about GitLab support.

Maybe change the sentence 
> This is a tool to help automate adding contributor acknowledgements according to the all-contributors specification.

to read
> This is a tool to help automate adding contributor acknowledgements according to the all-contributors specification for your GitHub or GitLab repository.

under *This solution* in the README file

**Comment by machour** [comment]
Outstanding work @tigermarques 👏 👏 👏 👏

File: src/repo/index.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+const gitlabAPI = require('./gitlab')
+
+const SUPPORTED_REPO_TYPES = {
+  github: {
+    value: 'github',
+    name: 'GitHub',
+    checkKey: 'login',
+    defaultHost: 'https://github.com',
+    linkToCommits: '<%= options.repoHost %>/<%= options.projectOwner %>/<%= options.projectName %>/commits?author=<%= contributor.login %>',
+    linkToIssues: '<%= options.repoHost %>/<%= options.projectOwner %>/<%= options.projectName %>/issues?q=author%3A<%= contributor.login %>',
```
Comment (kentcdodds) [inline src/repo/index.js]: Could we add `|| "https://github.com"` here like we have for gitlab? It makes sense to me. Could you open up another PR for this @tigermarques?

Comment (kentcdodds) [reply src/repo/index.js]: Actually, don't bother. I'll go ahead and do it really quick :)

**Pull request merged**